### PR TITLE
[DEV APPROVED] Add reCAPTCHA to Sign Up

### PR DESCRIPTION
**Pull request by willhall88**
**Target Process ticket**
https://moneyadviceservice.tpondemand.com/entity/8656

**Summary**
This PR adds reCAPTCHA to the registration page on Frontend.

---
This change is [Reviewable](https://reviewable.io/reviews/moneyadviceservice/frontend/1841)

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -35,6 +35,7 @@ gem 'nunes'
 gem 'opening_hours'
 gem 'psych', '>= 2.0.5' # https://www.ruby-lang.org/en/news/2014/03/29/heap-overflow-in-yaml-uri-escape-parsing-cve-2014-2525/
 gem 'redcarpet'
+gem "recaptcha", require: "recaptcha/rails"
```
Comment (tomas-stefano) [inline Gemfile]: Single quotes.

**Comment by tomas-stefano** [comment]
One minor comment but after the fix everything looks good to me  👍 👍

**Comment by margOnline** [comment]
nice work 👍

**Comment by Tr4pSt3R** [comment]
@willhall88 I run into the following issues from trying to run this.

The following files failed the Cucumber test. 

## Failing Scenarios
```
cucumber features/registration.feature:6 # Scenario: Registration when browsing MAS
cucumber features/registration.feature:14 # Scenario: Registration from a direct link
cucumber features/registration.feature:18 # Scenario: Register to participate in research
cucumber features/registration.feature:22 # Scenario: Default value for market research field
cucumber features/registration.feature:34 # Scenario Outline: Attempt to register with bad details, Examples (#1)
cucumber features/registration.feature:35 # Scenario Outline: Attempt to register with bad details, Examples (#2)
cucumber features/registration.feature:37 # Scenario: Attempt to register with existing account
cucumber features/registration.feature:43 # Scenario: Displaying sign up benefits when no session is set
cucumber features/registration.feature:47 # Scenario: Displaying sign up benefits when session is set
```

## Common Error
All the failures noticed in these files consistently pointed to the same issue which is given below. 

```
 Scenario: Displaying sign up benefits when session is set # features/registration.feature:47
    Given I have a registration title set in the session    # features/step_definitions/registration_steps.rb:108
    When I visit the registration page                      # features/step_definitions/registration_steps.rb:96
      No site key specified. (ActionView::Template::Error)
      ./app/views/devise/registrations/new.html.erb:129:in `block (2 levels) in _app_views_devise_registrations_new_html_erb___2346871571143813447_70153045580440'
      ./app/views/devise/registrations/new.html.erb:127:in `block in _app_views_devise_registrations_new_html_erb___2346871571143813447_70153045580440'
      ./app/views/devise/registrations/new.html.erb:16:in `_app_views_devise_registrations_new_html_erb___2346871571143813447_70153045580440'
      ./app/middleware/version_header.rb:13:in `call'
      ./app/middleware/route_probe.rb:19:in `call'
      ./app/middleware/override_head.rb:13:in `call'
      ./app/middleware/capture_request_id.rb:11:in `call'
      ./features/step_definitions/registration_steps.rb:97:in `/^I visit the registration page$/'
      ./features/support/env.rb:34:in `call'
      ./features/support/env.rb:34:in `block in <top (required)>'
      /opt/boxen/rbenv/versions/2.2.5/bin/bundle:23:in `load'
      /opt/boxen/rbenv/versions/2.2.5/bin/bundle:23:in `<main>'
      features/registration.feature:49:in `When I visit the registration page'
     Then I should not see the registration benefits
```

## What's The Issue?
This error occurs everywhere this line is called `<%= recaptcha_tags %>`. 

Upon investigation there are two required environmental variables which are missing. 

## Possible Solutions

### Solution 1 - Add a _Recaptcha_ Initialiser Config File
[See here](https://github.com/ambethia/recaptcha#alternative-api-key-setup) on how to add an alternative _Recaptcha_ API **site** and **secret** key. 

_NB_: I have tried this solution and it worked. All tests passed!

### Solution 2 - Add Missing Environmental Variables _Recaptcha_ 
This issue can also be solved by setting the env vars as follows:

```
export RECAPTCHA_SITE_KEY = '6Lc6BAAAAAAAAChqRbQZcn_yyyyyyyyyyyyyyyyy'
export RECAPTCHA_SECRET_KEY = '6Lc6BAAAAAAAAKN3DRm6VA_xxxxxxxxxxxxxxxxx'
```

I'm not sure why others did not get the errors I'm getting but it will be helpful to add the _Recaptcha_ keys to the [.env.sample](https://github.com/moneyadviceservice/frontend/blob/master/.env-example) file.

**Comment by willhall88** [comment]
yep my bad, I was planning on sorting documentation and adding the keys to scripts repo today. 
Keys are in our password manager ;) - I'll add the env vars expectation in the .env.sample file

File: Gemfile (hunk 1 above)
Comment (willhall88) [reply Gemfile]: good catch! thanks. I rubocopped some files. clearly missed this one!

File: config/locales/en.yml [hunk 2]
```diff
@@ -47,6 +47,10 @@ en:
     cy: Cymraeg
     en: English
 
+  recaptcha:
+    errors:
+      verification_failed: verification failed, please try again.
```
Comment (Tr4pSt3R) [inline config/locales/en.yml]: `verification failed, please try again.` --> 'Verification failed, please try again.'

The first letter 'V' should be capitalised. This is in lieu of the Welsh translation also being capitalised.

Comment (willhall88) [reply config/locales/en.yml]: money_advice_service
..maybe I need to go back to translation for this.. I was utilising the fact that in our validation messages we post the validation type first (Postcode, Email) and then the error, so you see the error like above ^ 

This is the error you see in Welsh. Capitalised to try and define the start of the sentence... 
y_gwasanaeth_cynghori_ariannol

Comment (willhall88) [reply config/locales/en.yml]: It's all kind of an awkward solution as reCAPTCHA itself does not support welsh language

Comment (Tr4pSt3R) [reply config/locales/en.yml]: @willhall88 what does UX have to say about this though? I do agree, it looks very awkward.

Comment (willhall88) [reply config/locales/en.yml]: who is this for a solution...? 
y_gwasanaeth_cynghori_ariannol

Comment (Tr4pSt3R) [reply config/locales/en.yml]: @willhall88 yeah that looks great. Should we capitalise the English locale then?

Comment (willhall88) [reply config/locales/en.yml]: I'd suggest no as this will still read as a sentence as seen above

Comment (Tr4pSt3R) [reply config/locales/en.yml]: @willhall88 ah right, that's true. The English version reads in a straigh-forward manner. Great work! Will

**Pull request merged**